music_assistant/server/providers/plex/__init__.py

```diff
@@ -0,0 +1,467 @@
+from music_assistant.server.helpers.tags import parse_tags
+from music_assistant.server.models import ProviderInstanceType
+from music_assistant.server.models.music_provider import MusicProvider
+
+CONF_AUTH_TOKEN = "token"
+CONF_SERVER_NAME = "server"
+CONF_LIBRARY_NAME = "library"
+
+
+async def setup(
+    mass: MusicAssistant, manifest: ProviderManifest, config: ProviderConfig
+) -> ProviderInstanceType:
+    """Initialize provider(instance) with given configuration."""
+    prov = PlexProvider(mass, manifest, config)
+    await prov.handle_setup()
+    return prov
+
+
+async def get_config_entries(
+    mass: MusicAssistant, manifest: ProviderManifest  # noqa: ARG001
+) -> tuple[ConfigEntry, ...]:
+    """Return Config entries to setup this provider."""
+    return (
+        ConfigEntry(
+            key=CONF_SERVER_NAME, type=ConfigEntryType.STRING, label="Server", required=True
+        ),
+        ConfigEntry(
+            key=CONF_LIBRARY_NAME, type=ConfigEntryType.STRING, label="Library", required=True
+        ),
+        ConfigEntry(
+            key=CONF_AUTH_TOKEN, type=ConfigEntryType.SECURE_STRING, label="Token", required=True
+        ),
+    )
+
+
+class PlexProvider(MusicProvider):
+    """Provider for a plex music library."""
+
+    _plex_server: PlexServer = None
+    _plex_library: PlexMusicSection = None
+
+    async def handle_setup(self) -> None:
+        """Set up the music provider by connecting to the server."""
+        if not self.config.get_value(CONF_AUTH_TOKEN):
+            raise LoginFailed("Invalid login credentials")
+
+        def connect():
+            plex_account = MyPlexAccount(token=self.config.get_value(CONF_AUTH_TOKEN))
+            return plex_account.resource(self.config.get_value(CONF_SERVER_NAME)).connect()
+
+        self._plex_server = await self._run_async(connect)
+        self._plex_library = await self._run_async(
+            self._plex_server.library.section, self.config.get_value(CONF_LIBRARY_NAME)
+        )
+
+    async def resolve_image(self, path: str) -> str | bytes | AsyncGenerator[bytes, None]:
+        """Return the full image URL including the auth token."""
+        return self._plex_server.url(path, True)
+
+    @property
+    def supported_features(self) -> tuple[ProviderFeature, ...]:
+        """Return a list of supported features."""
+        return (
+            ProviderFeature.LIBRARY_ARTISTS,
+            ProviderFeature.LIBRARY_ALBUMS,
+            ProviderFeature.LIBRARY_TRACKS,
+            ProviderFeature.LIBRARY_PLAYLISTS,
+            ProviderFeature.BROWSE,
+            ProviderFeature.SEARCH,
+            ProviderFeature.ARTIST_ALBUMS,
+        )
+
+    async def _run_async(self, call: Callable, *args, **kwargs):
+        return await self.mass.create_task(call, *args, **kwargs)
+
+    async def _get_data(self, key, cls=None):
+        return await self._run_async(self._plex_library.fetchItem, key, cls)
+
+    @alru_cache(maxsize=128, ttl=300)
```

Comment:
I'm not so sure about this. We should either switch all in-memory caching to lru_cache or keep with the music assistant cache, otherwise its hard to keep track of memory consumption. 

I'm leaning towards using the music assistant cache as that can also store longer TTL in a db but I'd like to hear your vision.

Reply:
for the record, the MA cache also has a decorator, just import use_cache from the cache module

Reply:
removed in favor of ItemMapping